README.md
# riveted [![Clojure CI](https://github.com/mudge/riveted/actions/workflows/clojure.yml/badge.svg)](https://github.com/mudge/riveted/actions/workflows/clojure.yml)

A Clojure library for the
[fast](http://vtd-xml.sourceforge.net/benchmark1.html) processing of XML with
[VTD-XML](http://vtd-xml.sourceforge.net), a [Virtual Token
Descriptor](http://vtd-xml.sf.net/VTD.html) XML parser.

It provides a more Clojure-like abstraction over VTD while still exposing the
power of its low-level interface.

## Installation

As riveted is available on [Clojars](https://clojars.org/riveted), add the
following to your [Leiningen](https://github.com/technomancy/leiningen)
dependencies:

```clojure
[riveted "0.2.0"]
```

## Compatibility

riveted is tested against Clojure 1.3, 1.4, 1.5.1, 1.6, 1.7, 1.8, 1.9, 1.10.0,
1.10.1, 1.10.2 and 1.10.3.

## API Documentation

The latest [riveted API documentation](http://mudge.name/riveted/) is
automatically generated with [Codox](https://github.com/weavejester/codox).

## Quick Start

For more details, see [Usage](#usage) below.

```clojure
(ns foo
  (:require [riveted.core :as vtd]))

(def nav (vtd/navigator (slurp "foo.xml")))

;; Navigating by direction and returning text content.
(-> nav vtd/first-child vtd/next-sibling vtd/text) ;=> "Foo"

;; Navigating by direction, restricted by element and returning attribute
;; value.
(-> nav (vtd/first-child :p) (attr :id)) ;=> "42"

;; Return the tag names of all children elements.
(->> nav vtd/children (map vtd/tag)) ;=> ("p" "a" "b")

;; Navigating by element name, regardless of location.
(-> nav (vtd/select :p) first vtd/text)

;; Navigating by XPath, returning all matches.
(map vtd/text (vtd/search nav "//author"))

;; Navigating by XPath, returning the first match.
(vtd/text (vtd/at nav "/article/title"))

;; Calling seq (or any function that uses seq such as first, second, nth,
;; last, etc.) on the navigator yields a sequence of all parsed tokens as
;; simple maps with a type and value entry.
(first nav) ;=> {:type :start-tag, :value "a"}
```

## Usage

NOTE: For your convenience, we created a sample project that you can use to experiment with riveted XML API.
See `examples/simple` project.

Once installed, you can include riveted into your desired namespace by
requiring `riveted.core` like so:

```clojure
(ns foo
  (:require [riveted.core :as vtd]))
```

The core data structure in riveted is the navigator: this represents both your
XML document and your current location within it. It can be interrogated for
the tag name, attributes and text value of any given element and also provides
the ability to move around the document.

Let's say we have a file called `foo.xml` with the following content:

```xml
<article>
  <title>Foo bar</title>
  <author id="1">
    <name>Robert Paulson</name>
    <name>Joe Bloggs</name>
  </author>
  <abstract>
    A <i>great</i> article all about things.
  </abstract>
</article>
```

Let's load this into an initial navigator with the `navigator` function,
passing it a UTF-8 encoded string of XML and then storing the result in the
[var](http://clojure.org/vars) `nav`:

```clojure
(def nav (vtd/navigator (slurp "foo.xml")))
```

If you already have your XML in a byte array, you can pass this directly to `navigator` instead of a UTF-8 string:

```clojure
(def nav (vtd/navigator my-byte-array))
```

`navigator` also takes an optional second argument to enable XML namespace
support which is disabled by default. We'll look at this
[later](#namespace-support) but, for now, we can process this document without
using namespaces.

Now that we have a navigator, we can navigate the document in several ways
(c.f. [VTD-XML's explanation of its different
views](http://vtd-xml.sourceforge.net/userGuide/3.html)):

* As a [cursor-based hierarchical view](#traversing-by-direction);
* Using [element selectors](#traversing-by-element-name);
* Using [XPath](#traversing-by-xpath);
* As a [flat view of tokens](#flat-view-of-tokens).

There is also a [mutable interface](#mutable-interface) for more constrained
memory usage.

### Traversing by direction

After parsing a document, the navigator's cursor is always at the root element
of our XML: for `foo.xml`, this means the `article` element. If we want to
retrieve the `title` and we know it's the first child of the article we can
simply use riveted's `first-child` function:

```clojure
(vtd/first-child nav)
```

This returns a new navigator with its cursor set to the `title` element. We
can check this by using the `text` and `tag` functions to return the text
content and tag name of the current cursor respectively:

```clojure
(vtd/text (vtd/first-child nav)) ;=> "Foo bar"
(vtd/tag (vtd/first-child nav))  ;=> "title"
```

If we then want to move to the `author` element, we can use the `next-sibling`
function in a similar way:

```clojure
(vtd/next-sibling (vtd/first-child nav))
```

It may be more readable to use Clojure's [threading macro,
`->`](http://clojuredocs.org/clojure_core/clojure.core/-%3E) when traversing
in multiple directions:

```clojure
(-> nav vtd/first-child vtd/next-sibling)
```

If we want to test an element for its attributes, we can use `attr?` like so:

```clojure
(-> nav vtd/first-child vtd/next-sibling (vtd/attr? :id)) ;=> true
```

We can then fetch the value of the attribute with `attr`:

```clojure
(-> nav vtd/first-child vtd/next-sibling (vtd/attr :id)) ;=> "1"

;; equivalent to:
(vtd/attr (vtd/next-sibling (vtd/first-child nav)) :id)
```

As well as `first-child` and `next-sibling`, you can move in one direction
with the following functions:

```clojure
(vtd/previous-sibling nav) ;=> move to the previous sibling element
(vtd/last-child nav)       ;=> move to the last child element
(vtd/parent nav)           ;=> move to the parent element
(vtd/root nav)             ;=> move to the root element
```

We can also test navigators to distinguish elements from the entire document:

```clojure
(-> nav vtd/first-child vtd/element?)   ;=> true
(-> nav vtd/parent vtd/document?)       ;=> true
(-> nav vtd/first-child vtd/attribute?) ;=> false
```

As we are positioned on the `author` element, we might now want to collect the
text values of the `name` elements within it. We could do this using the
directional functions above but riveted provides a `children` function to do
this for us:

```clojure
(->> nav vtd/first-child vtd/next-sibling vtd/children (map vtd/text))
;=> ("Robert Paulson" "Joe Bloggs")

;; or if you prefer not to use the threading macro:
(map vtd/text (vtd/children (vtd/next-sibling (vtd/first-child nav))))
```

Note that `children`, along with `next-siblings` and `previous-siblings`,
returns a lazy sequence of matching elements. They also take an optional
second argument which allows you to specify an element name which will
restrict results further.

For example, if you wanted to return the `author` element directly from the
original navigator, you could ask for the first `author` child like so:

```clojure
(-> nav (vtd/first-child :author))
```

Or ask the root for all child `author` elements:

```clojure
(-> nav (vtd/children :author)) ;=> a sequence of all author child elements
```

You can also get the full text content of a mixed-content node with `text`
which would be perfect for our `abstract` element:

```clojure
(-> nav (vtd/first-child :abstract) vtd/text)
;=> "A great article all about things."
```

If you want to retrieve the raw XML contents of a node, you can use `fragment`
to do so:

```clojure
(-> nav (vtd/first-child :abstract) vtd/fragment)
;=> "A <i>great</i> article all about things."
```

### Traversing by element name

If we'd rather not navigate a document in terms of directions, riveted also
provides a way to traverse XML by element names with `select`.

To continue our example from above, if we wanted to pull the `title` text, we
could ask the navigator for all `title` elements (regardless of location) like
so:

```clojure
(vtd/select nav :title)
```

As this is a lazy sequence, we can ask for the text of the first item like so:

```clojure
(-> nav (vtd/select :title) first vtd/text) ;=> "Foo bar"
```

Similarly, we can ask for the text value of all `name` elements like so:

```clojure
(map vtd/text (vtd/select nav :name)) ;=> ("Robert Paulson" "Joe Bloggs")
```

Note that this will return `name` elements *anywhere* in the document but we
could restrict its search by moving the navigator, perhaps using some of the
direction functions from above:

```clojure
(map vtd/text (-> nav (vtd/first-child :author) (vtd/select :name)))
;=> ("Robert Paulson" "Joe Bloggs")
```

Or perhaps with `select` itself:

```clojure
(map vtd/text (-> nav (vtd/select :author) first (vtd/select :name)))
;=> ("Robert Paulson" "Joe Bloggs")
```

Finally, we can return a lazy sequence of *all* elements by simply using a
wildcard match:

```clojure
(vtd/select nav "*")
```

### Traversing by XPath

The last way to traverse a document is to use XPath 1.0 with the `search`
function. Note that this is only used to navigate to elements (so it's not
possible to directly return attribute values with an XPath expression).

For example, to select all `name` elements:

```clojure
(vtd/search nav "//name")
```

If you are expecting only one match then you can use the `at` function to
return only one result:

```clojure
(vtd/at nav "/article/title")
```

If accessing attributes via XPath, you can use `text` to return the value of
the attribute:

```clojure
(text (vtd/at nav "/article/@id"))
```

### Namespace support

If you wish to use namespace-aware features, you will need to enable namespace
support when creating the initial navigator like so:

```clojure
(def ns-nav (vtd/navigator (slurp "namespaced.xml") true))
```

You can then pass a prefix and URL when using `search` and `at` like so:

```clojure
(vtd/search ns-nav "//ns1:name" "ns1" "http://purl.org/dc/elements/1.1/")
```

### Flat view of tokens

If you need lower level access to the parsed document, you can exploit the
fact that navigators implement [Clojure's `Seqable`
interface](http://clojure.org/sequences) and can be traversed as a flat
sequence much like a list or vector:

```clojure
(first nav)  ;=> {:type :start-tag, :value "article"}
(second nav) ;=> {:type :start-tag, :value "title"}
(nth nav 2)  ;=> {:type :character-data, :value "Foo bar"}
(nth nav 4)  ;=> {:type :attribute-name, :value "id"}
(seq nav)    ;=> the full sequence of tokens

;; Return all comments from a document.
(filter (comp #{:comment} :type) nav)
```

This gives you access to *all* tokens in the document including XML
declarations, doctypes, comments, processing instructions, etc. However, it is
a very low level of abstraction and if you only care about navigating
elements, it might be better to use a cursor-based view instead.

### Mutable interface

riveted also provides a mutable interface to
[VTDNav](http://vtd-xml.sourceforge.net/javadoc/com/ximpleware/VTDNav.html)
(much like Clojure's [transient](http://clojure.org/transients) data
structures) for lower-memory usage (at the cost of immutability):

```clojure
;; Create an initial navigator as per usual.
(def nav (navigator "<root><a>Foo</a><b>Bar</b></root>"))

;; Mutate nav to point to the a element.
(vtd/first-child! nav)

(vtd/text nav)
;=> "Foo"

;; Mutate nav to point to the b element.
(vtd/next-sibling! nav)

(vtd/text nav)
;=> "Bar"

;; Mutate nav to point to the a element again.
(vtd/previous-sibling! nav)

;; Mutate nav to point to the root element.
(vtd/parent! nav)

;; Mutate nav to point to the root of the document (regardless of location).
(vtd/root! nav)
```

In order to mitigate the problems with mutable state, it might be best to use
the above functions much like you would `transient`; viz. within the confines
of a function like so:

```clojure
(defn title [nav]
  (-> (vtd/root nav)                    ; Create a new navigator to the root
      (vtd/first-child! :front)         ; for mutation.
      (vtd/first-child! :article-meta)
      (vtd/first-child! :title-group)
      (vtd/first-child! :article-title)
      vtd/text))
```

In this way, only one extra navigator is created.

## Acknowledgements

[Andrew Diamond's `clj-vtd-xml`](https://github.com/diamondap/clj-vtd-xml) and
[Tim Williams' gist](https://gist.github.com/willtim/822769) are existing
interfaces to VTD-XML from Clojure that were great sources of inspiration.

[Dave Ray's `seesaw`](https://github.com/daveray/seesaw) set the standard for
helpful docstrings.

Clojure's
[`core.clj`](https://github.com/clojure/clojure/blob/master/src/clj/clojure/core.clj)
provided fascinating reading, particularly regarding the use of `:inline`
metadata.

Thanks to [Heikki Hämäläinen](https://github.com/hjhamala) for contributing a
character encoding fix for Windows users.

Thanks to [Eugen Stan](https://github.com/ieugen) for suggesting that
`navigator` should also accept byte arrays as well as UTF-8 strings.

## License

Copyright © 2013-2022 Paul Mucur.

Distributed under the Eclipse Public License, the same as Clojure.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2022-02-15
### Changed
- Added support for passing a byte array directly to `navigator` rather than only
  accepting UTF-8 strings (thanks to Eugen Stan for the suggestion)
- Passing an invalid value to `navigator` (e.g. `nil`) will now throw an
  `IllegalArgumentException` rather than a `NullPointerException`

## [0.1.2] - 2022-02-11
### Changed
- Upgrade underlying VTD-XML dependency to 2.13.4 (thanks to Eugen Stan)

## [0.1.1] - 2018-01-29
### Fixed
- Fixed exception when calling fragment on a self-closing tag

## [0.1.0] - 2017-06-27
### Fixed
- Explicitly set the character set to UTF-8 when reading XML

### Changed
- Upgrade underlying VTD-XML dependency to 2.13

## [0.0.9] - 2013-09-05
### Added
- Add ability to return the text of an attribute using `text`
- Add ability to check if the navigator is pointed at an attribute with `attribute?`

## [0.0.8] - 2013-05-02
### Added
- Gracefully handle `nil` across the library to make threading easier

## [0.0.7] - 2013-03-29
### Added
- Enrich navigators so that are now sequential, seqable and countable data structures

## [0.0.6] - 2013-03-29
### Added
- Added `select` for selecting elements by name or wildcard

## [0.0.5] - 2013-03-28
### Added
- Added transient interface via `root!`, `parent!`, `next-sibling!`, `previous-sibling!`, `first-child!` and `last-child!`

## [0.0.4] - 2013-03-25
### Added
- Added `attr?` for testing the existence of attributes
- Added ability to pass optional element names to `first-child`, `last-child`,
  `next-sibling`, `previous-sibling`, `siblings` and `children`

## [0.0.3] - 2013-03-24
### Added
- Added support for XML namespaces when searching

### Changed
- Replaced public `token-type` with `element?` and `document?` functions

## [0.0.2] - 2013-03-24
### Added
- Add ability to fetch previous sibling
- Add ability to fetch token type for an element

### Fixed
- Fix fetching all siblings for a navigator, both previous and next

## [0.0.1] - 2013-03-24
### Added
- First stable version of riveted

[0.2.0]: https://github.com/mudge/riveted/releases/tag/v0.2.0
[0.1.2]: https://github.com/mudge/riveted/releases/tag/v0.1.2
[0.1.1]: https://github.com/mudge/riveted/releases/tag/v0.1.1
[0.1.0]: https://github.com/mudge/riveted/releases/tag/v0.1.0
[0.0.9]: https://github.com/mudge/riveted/releases/tag/v0.0.9
[0.0.8]: https://github.com/mudge/riveted/releases/tag/v0.0.8
[0.0.7]: https://github.com/mudge/riveted/releases/tag/v0.0.7
[0.0.6]: https://github.com/mudge/riveted/releases/tag/v0.0.6
[0.0.5]: https://github.com/mudge/riveted/releases/tag/v0.0.5
[0.0.4]: https://github.com/mudge/riveted/releases/tag/v0.0.4
[0.0.3]: https://github.com/mudge/riveted/releases/tag/v0.0.3
[0.0.2]: https://github.com/mudge/riveted/releases/tag/v0.0.2
[0.0.1]: https://github.com/mudge/riveted/releases/tag/v0.0.1


test/riveted/core_test.clj
(ns riveted.core-test
  (:use midje.sweet riveted.core)
  (:import com.ximpleware.VTDNav))

;;; Test data.

(def xml "<root><!--Hello--><basic-title>Foo</basic-title><complex-title id=\"42\" empty=\"\"><i>Foo</i> woo <b>moo</b></complex-title><i>Bar</i><foo/></root>")
(def ns-xml "<root xmlns:dc=\"http://purl.org/dc/elements/1.1/\"><dc:name>Bob</dc:name></root>")

(def nav (navigator xml false))
(def ns-nav (navigator ns-xml true))

;;; Custom checkers to simplify testing.

(defn nav? [actual] (instance? riveted.core.Navigator actual))
(defn tag? [tag-name] (fn [actual] (= tag-name (tag actual))))
(defn tags? [& tag-names] (fn [actual] (= tag-names (map tag actual))))
(def root? (tag? "root"))

(fact "navigator returns a VTD navigator for a byte array."
  (navigator (.getBytes "<a></a>" "UTF-8")) => nav?
  (navigator (.getBytes "<a></a>" "ISO-8859-1")) => nav?)

(fact "navigator returns a VTD navigator for a UTF-8 string."
  (navigator "<a></a>") => nav?)

(fact "navigator raises an IllegalArgumentException nil if given nil."
  (navigator nil) => (throws IllegalArgumentException))

(fact "search returns a sequence of matching navigators for a given XPath."
  (search nav "/root/basic-title") => (one-of nav?)
  (search nav "//i") => (two-of nav?)
  (search nav "/missing") => empty?
  (search ns-nav "/root/foo:name" "foo" "http://purl.org/dc/elements/1.1/") => (one-of nav?)
  (search nil "/foo") => empty?)

(fact "at returns the first matching navigator for a given XPath."
  (at nav "/root/basic-title") => nav?
  (at nav "/missing") => nil?
  (at ns-nav "/root/foo:name" "foo" "http://purl.org/dc/elements/1.1/") => nav?
  (at nil "/foo") => nil?)

(fact "select returns navigators for all matching elements."
  (select nav "*") => (tags? "root" "basic-title" "complex-title" "i" "b" "i" "foo")
  (select nav "i") => (two-of (tag? "i"))
  (select nav :i) => (two-of (tag? "i"))
  (select (at nav "/root/complex-title") "*") => (tags? "complex-title" "i" "b")
  (select nav "missing") => empty?
  (select nil "foo") => empty?)

(fact "text returns the text nodes descending from a navigator."
  (text (at nav "/root/basic-title")) => "Foo"
  (text (at nav "/root/complex-title")) => "Foo woo moo"
  (text (at nav "/root/foo")) => nil?
  (text nil) => nil?)

(fact "fragment returns the content of the given navigator as an XML fragment."
  (fragment (at nav "/root/complex-title")) => "<i>Foo</i> woo <b>moo</b>"
  (fragment (at nav "/root/basic-title")) => "Foo"
  (fragment (at nav "/root/foo")) => ""
  (fragment nil) => nil?)

(fact "attr returns the value of the given attribute"
  (attr (at nav "/root/complex-title") :id) => "42"
  (attr (at nav "/root/complex-title") "id") => "42"
  (attr (at nav "/root/complex-title") :missing) => nil?
  (attr nil :foo) => nil?)

(fact "tag returns the current element name of the given navigator."
  (tag (root nav)) => "root"
  (tag (at nav "/root/complex-title")) => "complex-title"
  (tag nil) => nil?)

(fact "document? returns true if the navigator is set to the document."
  (parent (root nav)) => document?
  (root nav) =not=> document?
  (document? nil) => false)

(fact "element? returns true if the navigator is set to an element."
  (root nav) => element?
  (parent (root nav)) =not=> element?
  (element? nil) => false)

(fact "parent returns a navigator for the parent of the current navigator."
  (parent (at nav "/root/basic-title")) => nav?
  (parent nil) => nil?)

(fact "parent returns the document as the parent of the root."
  (parent (root nav)) => document?)

(fact "parent returns nil for the parent of the document."
  (parent (parent (root nav))) => nil?)

(fact "root returns a navigator for the root element."
  (root (at nav "/root/complex-title/i")) => nav?
  (root (at nav "/root/complex-title/i")) => root?)

(fact "first-child returns a navigator for the first child element."
  (first-child (at nav "/root/complex-title")) => nav?
  (first-child (at nav "/root/complex-title")) => (tag? "i")
  (first-child (at nav "/root/foo")) => nil?)

(fact "first-child takes an optional element name."
  (first-child (root nav) :complex-title) => (tag? "complex-title")
  (first-child (root nav) "complex-title") => (tag? "complex-title")
  (first-child (root nav) :missing) => nil?)

(fact "last-child returns a navigator for the last child element."
  (last-child (at nav "/root/complex-title")) => (tag? "b")
  (last-child (at nav "/root/foo")) => nil?)

(fact "last-child takes an optional element name."
  (last-child (root nav) :complex-title) => (tag? "complex-title")
  (last-child (root nav) "complex-title") => (tag? "complex-title")
  (last-child (root nav) :missing) => nil?)

(fact "next-sibling returns a navigator for the next sibling element."
  (next-sibling (at nav "/root/basic-title")) => (tag? "complex-title")
  (next-sibling (at nav "/root/complex-title")) => (tag? "i")
  (next-sibling (at nav "/root/foo")) => nil?)

(fact "next-sibling takes an optional element name."
  (next-sibling (at nav "/root/basic-title") :i) => (tag? "i")
  (next-sibling (at nav "/root/basic-title") "i") => (tag? "i")
  (next-sibling (at nav "/root/basic-title") :missing) => nil?)

(fact "previous-sibling returns a navigator for the previous sibling element."
  (previous-sibling (at nav "/root/complex-title")) => (tag? "basic-title")
  (previous-sibling (at nav "/root/foo")) => (tag? "i")
  (previous-sibling (at nav "/root/basic-title")) => nil?)

(fact "previous-sibling takes an optional element name."
  (previous-sibling (at nav "/root/i") :basic-title) => (tag? "basic-title")
  (previous-sibling (at nav "/root/i") "basic-title") => (tag? "basic-title")
  (previous-sibling (at nav "/root/i") :missing) => nil?)

(fact "next-siblings returns navigators for all next sibling elements."
  (next-siblings (at nav "/root/basic-title")) => (tags? "complex-title" "i"
                                                         "foo")
  (next-siblings (at nav "/root/foo")) => empty?)

(fact "next-siblings takes an optional element name."
  (next-siblings (at nav "/root/basic-title") :i) => (tags? "i")
  (next-siblings (at nav "/root/basic-title") "i") => (tags? "i")
  (next-siblings (at nav "/root/basic-title") :missing) => empty?)

(fact "previous-siblings returns navigators for all previous sibling elements."
  (previous-siblings (at nav "/root/foo")) => (tags? "i" "complex-title"
                                                     "basic-title")
  (previous-siblings (at nav "/root/basic-title")) => empty?)

(fact "previous-siblings takes an optional element name."
  (previous-siblings (at nav "/root/foo") :i) => (tags? "i")
  (previous-siblings (at nav "/root/foo") "i") => (tags? "i")
  (previous-siblings (at nav "/root/foo") :missing) => empty?)

(fact "siblings returns navigators for all sibling elements."
  (siblings (at nav "/root/basic-title")) => (tags? "complex-title" "i" "foo")
  (siblings (at nav "/root/complex-title")) => (tags? "basic-title" "i" "foo")
  (siblings (at nav "/root/i")) => (tags? "basic-title" "complex-title" "foo")
  (siblings (root nav)) => empty?)

(fact "siblings takes an optional element name."
  (siblings (at nav "/root/basic-title")
            :complex-title) => (tags? "complex-title")
  (siblings (at nav "/root/basic-title")
            "complex-title") => (tags? "complex-title")
  (siblings (at nav "/root/basic-title") :missing) => empty?)

(fact "children returns navigators for all children elements."
  (children (root nav)) => (tags? "basic-title" "complex-title" "i" "foo")
  (children (at nav "/root/foo")) => empty?)

(fact "children takes an optional element name."
  (children (root nav) :complex-title) => (tags? "complex-title")
  (children (root nav) "complex-title") => (tags? "complex-title")
  (children (root nav) :missing) => empty?)

(fact "attr? returns whether or not a given attribute exists."
  (attr? (at nav "/root/complex-title") :id) => true
  (attr? (at nav "/root/complex-title") "id") => true
  (attr? (at nav "/root/complex-title") :missing) => false)

;;; Mutable tests. Note that the order of these facts is critical.

(def nav! (navigator "<root><child><name>Foo</name><age>42</age></child><bro>Bar</bro></root>" false))

(fact "root! moves the given navigator to the root."
  (root! nav!) => (tag? "root"))

(fact "first-child! moves the given navigator to the first child."
  (first-child! nav!) => (tag? "child"))

(fact "first-child! takes an optional element name."
  (first-child! nav! :name) => (tag? "name")
  (first-child! nav! :missing) => nil?)

(fact "parent! moves the given navigator to the parent element."
  (parent! nav!) => (tag? "child"))

(fact "next-sibling! moves to the given navigator to the next sibling element."
  (next-sibling! nav!) => (tag? "bro")
  (next-sibling! nav!) => nil?)

(fact "next-sibling! takes an optional element name."
  (next-sibling! (previous-sibling! nav!) :bro) => (tag? "bro")
  (next-sibling! (previous-sibling! nav!) "bro") => (tag? "bro")
  (next-sibling! nav! :missing) => nil?)

(fact "previous-sibling! moves the given navigator to the previous sibling."
  (previous-sibling! nav!) => (tag? "child")
  (previous-sibling! nav!) => nil?)

(fact "previous-sibling! takes an optional element name."
  (previous-sibling! (next-sibling! nav!) :child) => (tag? "child")
  (previous-sibling! (next-sibling! nav!) "child") => (tag? "child")
  (previous-sibling! nav! :missing) => nil?)

(fact "last-child! moves the given navigator to the last child."
  (last-child! nav!) => (tag? "age"))

(fact "navigators are sequential."
  nav => sequential?)

(fact "navigators are counted."
  nav => counted?
  (count nav) => 18)

(fact "navigators expose all internal tokens as a seq."
  (first nav)  => {:type :start-tag, :value "root"}
  (second nav) => {:type :comment, :value "Hello"}
  (nth nav 2)  => {:type :start-tag, :value "basic-title"}
  (nth nav 3)  => {:type :character-data, :value "Foo"}
  (nth nav 5)  => {:type :attribute-name, :value "id"}
  (nth nav 6)  => {:type :attribute-value, :value "42"})

(fact "navigators not at the root, seq the remaining nodes."
  (first (first-child nav :complex-title)) => {:type :start-tag,
                                               :value "complex-title"}
  (last (first-child nav :complex-title)) => {:type :start-tag,
                                              :value "foo"})

(fact "navigators can safely be threaded even with nils."
  (-> nav (first-child :missing) (last-child :missing) text) => nil?)

(fact "attribute? returns true if the navigator is set to an attribute."
  (at nav "/root/complex-title/@id") => attribute?
  (at nav "/root/complex-title") =not=> attribute?
  (root nav) =not=> attribute?)

(fact "attribute values can be retrieved with text."
  (text (at nav "/root/complex-title/@id")) => "42"
  (text (at nav "/root/complex-title/@empty")) => "")

(fact "attribute names can be retrieved with tag."
  (tag (at nav "/root/complex-title/@id")) => "id"
  (tag (at nav "/root/complex-title/@empty")) => "empty")



examples/simple/src/foo.clj
(ns foo
  (:require [riveted.core :as vtd]))

;; https://github.com/mudge/riveted#usage
(comment
  (def nav (vtd/navigator (slurp "resources/foo.xml")))

;; Navigating by direction and returning text content.
  (-> nav vtd/first-child vtd/next-sibling vtd/text)
   ;=> "Foo"

;; Navigating by direction, restricted by element and returning attribute
;; value.
  (-> nav (vtd/first-child :p) (vtd/attr :id))
   ;=> "42"

;; Return the tag names of all children elements.
  (->> nav vtd/children (map vtd/tag))
   ;=> ("p" "a" "b")

;; Navigating by element name, regardless of location.
  (-> nav (vtd/select :p) first vtd/text)

;; Navigating by XPath, returning all matches.
  (map vtd/text (vtd/search nav "//author"))

;; Navigating by XPath, returning the first match.
  (vtd/text (vtd/at nav "/article/title"))

;; Calling seq (or any function that uses seq such as first, second, nth,
;; last, etc.) on the navigator yields a sequence of all parsed tokens as
;; simple maps with a type and value entry.
  (first nav) ;=> {:type :start-tag, :value "a"}
  )

(comment
;; https://github.com/mudge/riveted#mutable-interface

;; Create an initial navigator as per usual.
  (def nav (vtd/navigator "<root><a>Foo</a><b>Bar</b></root>"))

;; Mutate nav to point to the a element.
  (vtd/first-child! nav)

  (vtd/text nav)
;=> "Foo"

;; Mutate nav to point to the b element.
  (vtd/next-sibling! nav)

  (vtd/text nav)
;=> "Bar"

;; Mutate nav to point to the a element again.
  (vtd/previous-sibling! nav)

;; Mutate nav to point to the root element.
  (vtd/parent! nav)

;; Mutate nav to point to the root of the document (regardless of location).
  (vtd/root! nav))

project.clj
(defproject riveted "0.2.0"
  :description "A Clojure library for the fast processing of XML with VTD-XML."
  :url "https://github.com/mudge/riveted"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[com.ximpleware/vtd-xml "2.13.4"]]
  :profiles {:dev {:dependencies [[midje "1.10.5"]]
                   :plugins [[lein-midje "3.2.2"]
                             [lein-codox "0.10.8"]]}
             :1.3 {:dependencies [[midje "1.6.3"] [org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [[midje "1.6.3"] [org.clojure/clojure "1.4.0"]]}
             :1.5.1 {:dependencies [[midje "1.6.3"] [org.clojure/clojure "1.5.1"]]}
             :1.6 {:dependencies [[midje "1.6.3"] [org.clojure/clojure "1.6.0"]]}
             :1.7 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.8.0"]]}
             :1.9 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.9.0"]]}
             :1.10.0 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.10.0"]]}
             :1.10.1 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.10.1"]]}
             :1.10.2 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.10.2"]]}
             :1.10.3 {:dependencies [[midje "1.10.5"] [org.clojure/clojure "1.10.3"]]}}
  :aliases {"all" ["with-profile" "dev,1.3:dev,1.4:dev,1.5.1:dev,1.6:dev,1.7:dev,1.8:dev,1.9:dev,1.10.0:dev,1.10.1:dev,1.10.2:dev,1.10.3"]}
  :codox {:src-dir-uri "https://github.com/mudge/riveted/blob/main"
          :src-linenum-anchor-prefix "L"})
